hoomd/filter/custom.py

```diff
@@ -0,0 +1,54 @@
+"""Contains a class for custom particle filters in Python."""
+from abc import ABC, abstractmethod
+from collections.abc import Hashable, Callable
+
+class CustomFilter(Hashable, Callable):
+    """Abstract base class for custom particle filters.
+
+    The class allows the definition of particle filters in Python (see
+    `hoomd.filter.ParticleFilter`.
+    """
+    @abstractmethod
+    def __call__(self, state):
+        """Return the local particle tags that match the filter.
+
+        Returns the tags that are local to an MPI rank that match the particle
+        filter. Tag numbers in a `hoomd.Snapshot` object are just their index.
+
+        Note:
+            The exact requirements for the tags returned by custom filters on
+            each MPI rank is that the set union of the returned arrays from each
+            MPI rank be all particles that match the filter. For general use, it
+            is recommended, however, that each rank only return the tags for
+            particles that are in the local MPI rank (excluding ghost
+            particles). This is preferable for ease of use with local snapshots
+            by avoiding accidentally attempting to access invalid array indices
+            from tags outside of the MPI rank.
+
+        Args:
+            state (`hoomd.State`):
+                The simulation state to return the filtered tags from.
+
+        Returns:
+            (N,) `numpy.ndarray` of `numpy.uint64`:
+                An array of MPI local tags that match the filter.
+        """
+        pass
+
+    @abstractmethod
+    def __hash__(self):
+        """A hashed value to represent this instance.
+
+        This is necessary to allow for proper caching of filter tags internally
+        in HOOMD-blue.
+        """
+        pass
```

Comment:
Better yet, just inherit from `Hashable`.

https://docs.python.org/3/library/collections.abc.html

Reply:
The class is already a subclass of both `Hashable` and `Callable`. I assumed that @b-butler overrode these methods with `abstractmethod`s so that he could document the signature and behavior of these methods, which I think can be appropriate for `__call__` since it can pretty much have an arbitrary signature and is part of the public API, but is unnecessary for `__hash__` which has a well-defined signature. My recommendation was to document _why_ this class is hashable as part of the class docstring rather than overriding the method (which is superfluous due to inheriting from `Hashable`) and adding a docstring.

Reply:
@bdice Vyas is correct, I can remove `__hash__` though and put it in the docstring.